src/image.rs

```diff
@@ -0,0 +1,1192 @@
+        Ok(Self::allocate(
+            width,
+            height,
+            format.scanline_pad.try_into()?,
+            depth,
+            format.bits_per_pixel.try_into()?,
+            setup.image_byte_order,
+        ))
+    }
+
+    /// The stride is the number of bytes that each row of pixel data occupies in memory.
+    fn stride(&self) -> usize {
+        compute_stride(self.width, self.bits_per_pixel, self.scanline_pad)
+    }
+
+    /// Get an image from the X11 server.
+    ///
+    /// This function sends a [`GetImage`](crate::protocol::xproto::GetImage) request, waits for
+    /// its response and wraps it in a new `Image`.
+    ///
+    /// The returned image contains the rectangle with top left corner `(x, y)` and size `(width,
+    /// height)` of the given `drawable`.
+    pub fn get(
+        conn: &impl Connection,
+        drawable: Drawable,
+        x: i16,
+        y: i16,
+        width: u16,
+        height: u16,
+    ) -> Result<Self, ReplyError> {
+        let reply = GetImageRequest {
+            drawable,
+            x,
+            y,
+            width,
+            height,
+            format: ImageFormat::ZPixmap,
+            plane_mask: !0,
+        }
+        .send(conn)?
+        .reply()?;
+        Ok(Self::get_from_reply(&conn.setup(), width, height, reply)?)
+    }
+
+    /// Construct an `Image` from a `GetImageReply`.
+    ///
+    /// This function takes a `GetImageReply` and wraps it in an `Image`. The given `width` and
+    /// `height` describe the corresponding values of the `GetImage` request that was used to get
+    /// the `GetImageReply`.
+    pub fn get_from_reply(
+        setup: &Setup,
+        width: u16,
+        height: u16,
+        reply: GetImageReply,
+    ) -> Result<Self, ParseError> {
+        let format = find_format(setup, reply.depth)?;
+        Self::new(
+            width,
+            height,
+            format.scanline_pad.try_into()?,
+            reply.depth,
+            format.bits_per_pixel.try_into()?,
+            ImageOrder::MSBFirst,
+            Cow::Owned(reply.data),
+        )
+    }
+
+    /// Put an image to the X11 server.
+    ///
+    /// This function sends a [`PutImage`](crate::protocol::xproto::PutImage) request. This will
+    /// upload this image to the given `drawable` to position `(dst_x, dst_y)`.
+    pub fn put<'c, Conn: Connection>(
+        &self,
+        conn: &'c Conn,
+        drawable: Drawable,
+        gc: Gcontext,
+        dst_x: i16,
+        dst_y: i16,
+    ) -> Result<VoidCookie<'c, Conn>, ConnectionError> {
+        PutImageRequest {
+            format: ImageFormat::ZPixmap,
+            drawable,
+            gc,
+            width: self.width,
+            height: self.height,
+            dst_x,
+            dst_y,
+            left_pad: 0, // Must always be 0 for ZPixmap
+            depth: self.depth,
+            data: Cow::Borrowed(&self.data),
+        }
+        .send(conn)
+    }
```

Comment:
I believe that at some point this should be improved to avoid generating too large requests.

Reply:
Urgh. Good idea, but this kills the return value, since the function can only return one `VoidCookie`:
```diff
diff --git a/src/image.rs b/src/image.rs
index 957d89b..d2a2850 100644
--- a/src/image.rs
+++ b/src/image.rs
@@ -25,7 +25,6 @@ use std::borrow::Cow;
 use std::convert::{TryFrom, TryInto};
 
 use crate::connection::Connection;
-use crate::cookie::VoidCookie;
 use crate::errors::{ConnectionError, ParseError, ReplyError};
 use crate::protocol::xproto::{
     Drawable, Format, Gcontext, GetImageReply, GetImageRequest, ImageFormat, ImageOrder,
@@ -710,20 +709,37 @@ impl<'a> Image<'a> {
         gc: Gcontext,
         dst_x: i16,
         dst_y: i16,
-    ) -> Result<VoidCookie<'c, Conn>, ConnectionError> {
-        PutImageRequest {
-            format: ImageFormat::ZPixmap,
-            drawable,
-            gc,
-            width: self.width,
-            height: self.height,
-            dst_x,
-            dst_y,
-            left_pad: 0, // Must always be 0 for ZPixmap
-            depth: self.depth,
-            data: Cow::Borrowed(&self.data),
+    ) -> Result<(), ConnectionError> {
+        // Upload the image without exceeding the server's maximum request size
+        let max_bytes = conn.maximum_request_bytes();
+        let put_image_header = 24;
+        let stride = self.stride();
+        let lines_per_request = (max_bytes - put_image_header) / stride;
+        let lines_per_request = lines_per_request.try_into().unwrap_or(u16::max_value());
+
+        let (mut y_offset, mut byte_offset) = (0, 0);
+        while y_offset < self.height {
+            let next_lines = lines_per_request.min(self.height - y_offset);
+            let next_byte_offset = byte_offset + usize::from(next_lines) * stride;
+            let data = &self.data[byte_offset..next_byte_offset];
+            let request = PutImageRequest {
+                format: ImageFormat::ZPixmap,
+                drawable,
+                gc,
+                width: self.width,
+                height: next_lines,
+                dst_x,
+                dst_y,
+                left_pad: 0, // Must always be 0 for ZPixmap
+                depth: self.depth,
+                data: Cow::Borrowed(&data),
+            };
+            let _cookie = request.send(conn);
+
+            y_offset += next_lines;
+            byte_offset = next_byte_offset;
         }
-        .send(conn)
+        Ok(())
     }
 
     /// Convert this image into the format specified by the other parameters.
```

Reply:
How about returning a `Vec<VoidCookie>` here? That seems to be the simplest way out. How bad would that be?

Reply:
I think that would be the best option. I don't think needing a loop to check the requests is that bad.

Reply:
Okay, done. I also added the `assert!` to the code that I mentioned in an edit in my comment above.